File: 0214-shortest-palindrome/0214-shortest-palindrome.cpp
class Solution {
public:
    
    vector< int> kmp( string s){
        
        int n=s.size();
        vector< int> k(n,0); // longest proper prefix .
        int i=1,l=0;
        
        while( i<n){
            if(s[l]==s[i]){
                l++;
                k[i]=l;
                i++;
            }
            else{
                if( l!=0){
                    l=k[l-1];
                }
                else{
                    i++;
                }
            }
        }
        return k;
        
    }
    string shortestPalindrome(string s) {
        
        string x=s;
        reverse(x.begin(),x.end());
        vector< int> a=kmp(s+"#"+x);
        
        // ( ( substr(l).reverse+longest pallindrome prefix i.e from start(i=0))
        x=s.substr(a.back());
        reverse(x.begin(),x.end());
        return x+s;
    }
};

File: 0214-shortest-palindrome/0214-shortest-palindrome.java
class Solution {
    
    public static int kmp( String s){
        
        int n=s.length(),i=1,l=0;
        int[] k=new int[n]; // longest proper prefix .
        k[0]=0;
        
        while( i<n){
            if(s.charAt(l)==s.charAt(i)){
                k[i++]=++l;
            }
            else{
                if( l!=0)l=k[l-1];
                else i++;
            }
        }
        return k[n-1];
    }
    public String shortestPalindrome(String s) {
        
        String x=s+"#"+ new StringBuilder(s).reverse().toString();
        return new StringBuilder(s.substring(kmp(x))).reverse().toString()+s;
        
    }
}